pex/installer.py

```diff
@@ -76,43 +56,37 @@ def _setup_command(self):
     """the setup command-line to run, to be implemented by subclasses."""
     raise NotImplementedError
 
-  def _postprocess(self):
-    """a post-processing function to run following setup.py invocation."""
+  @contextmanager
+  def bootstrap_script(self):
+    with third_party.isolated() as root:
+      yield """
+import sys
+sys.path.insert(0, {root!r})
+
+# Expose vendored mixin path_items (setuptools, wheel, etc.) directly to the package's setup.py.
+from pex import third_party
+third_party.install(root={root!r}, expose={mixins!r})
 
-  @property
-  def capability(self):
-    """returns the list of requirements for the interpreter to run this installer."""
-    return list(self.mixins().values())
+__file__ = 'setup.py'
+with open(__file__, 'rb') as fp:
+  ast = compile(fp.read(), __file__, 'exec')
 
-  @property
-  def bootstrap_script(self):
-    bootstrap_sys_paths = []
-    bootstrap_modules = []
-    for module, requirement in self.mixins().items():
-      path = self._interpreter.get_location(requirement)
-      if not path:
-        assert not self._strict  # This should be caught by validation
-        continue
-      bootstrap_sys_paths.append(self.SETUP_BOOTSTRAP_PYPATH % {'path': path})
-      bootstrap_modules.append(self.SETUP_BOOTSTRAP_MODULE % {'module': module})
-    return '\n'.join(
-      [self.SETUP_BOOTSTRAP_HEADER] +
-      bootstrap_sys_paths +
-      bootstrap_modules +
-      [self.SETUP_BOOTSTRAP_FOOTER]
-    )
+sys.argv[0] = __file__
+exec(ast)
```

Comment:
The high-level context is that we're executing the `setup.py` of an sdist here and attempting to control its environment - injecting setuptools and (optionally) wheel into sys.path so that `bdist_egg` and (optionally) bdist_wheel work. As such we can't `python setup.py bdist_wheel` against their setup.py directly since we need to inject `sys.path.insert(...)` modifications (granted - another approach would be to use `PYTHONPATH` and not pass `-E` to the interpreter - but I'm sticking with the existing technique and better isolation here).

As to the exec - I'm not exactly sure why it was done this way. Declaring `__file__` before `exec` suffices to inject `__file__` into the exec'd code which I think could have been the only point - let the 3rdparty `setup.py` think its `setup.py` even though it's being slurped in as a subset of a virtual setup.py pumped through stdin.

Comment added.